toc/rfc/rfc-draft-system-cnb.md

```diff
@@ -0,0 +1,27 @@
+# Meta
+[meta]: #meta
+- Name: System Cloud Native Buildpacks
+- Start Date: 30.04.2024
+- Author(s): @c0d1ngm0nk3y, @pbusko, @nicolasbender, @modulo11
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (fill in with PR link after you submit it)
+- Predecessor: https://github.com/cloudfoundry/community/pull/796
+
+## Summary
+
+This RFC proposes extending the Cloud Foundry functionality to introduce and support system Cloud Native Buildpacks (CNBs). It will allow users to initiate the `cf push --lifecycle cnb` command without providing an explicit list of buildpacks, relying instead on the list of system buildpacks and the built-in auto-detection mechanism.
+
+## Problem
+
+Despite the recent integration of CNBs in the Cloud Foundry platform, the feature still has limitations that could hinder user experience and efficiency. Currently, only custom buildpacks are supported. Users have to manually provide a list of buildpacks (at least one) for their application to run with the `cnb` lifecycle type.
+
+## Proposal
+
+- Introduce a new `lifecycle` column to the [buildpack object](https://v3-apidocs.cloudfoundry.org/version/3.164.0/index.html#the-buildpack-object) with ENUM values `cnb` or `buildpack`, with `buildpack` being the default.
+- Add new parameter `lifecycle` to the [`POST /v3/buildpacks`](https://v3-apidocs.cloudfoundry.org/version/3.164.0/index.html#create-a-buildpack) API endpoint, with the default value being inferred from the `default_app_lifecycle` setting within the Cloud Controller config.
+- Cloud Native Buildpacks MUST be stored in format of a gzipped OCI tarball. The layout of the tarball is outlined in the [OCI Image Layout Specification](https://github.com/opencontainers/image-spec/blob/main/image-layout.md)
+- Add support for gzipped OCI tarballs when uploading a buildpacks using [`POST /v3/buildpacks/:guid/upload`](https://v3-apidocs.cloudfoundry.org/version/3.164.0/index.html#upload-buildpack-bits) API endpoint
+- `cf push --lifecycle cnb` command without an explicit list of buildpacks MUST perform auto-detection using system Cloud Native Buildpacks based on their priority.
+- `cf push [--lifecycle buildpack]` command without specifying an explicit list of buildpacks MUST perform auto-detection using the system buildpacks based on their priority. The type of the buildpacks is based on the `default_app_lifecycle` setting within the Cloud Controller config.
```

Comment:
This RFC does not introduce `default_app_lifecycle` setting, this setting is already part of and is in use by Cloud Controller:

https://github.com/cloudfoundry/capi-release/blob/6a8ac6f03f8a79b8f6ed2556e9c8b469a12634ea/jobs/cloud_controller_ng/templates/cloud_controller_ng.yml.erb#L540

Reply:
Hi @pbusko,

the comment made by @stephanme is about splitting:
```
- `cf push [--lifecycle buildpack]` command without specifying an explicit list of buildpacks MUST perform auto-detection using the system buildpacks based on their priority. The type of the buildpacks is based on the `default_app_lifecycle` setting within the Cloud Controller config.
```
into `cf push` and `cf push --lifecycle buildpack` sections because the `cf push --lifecycle buildpack` option is not explained at all with the current formulation.

Reply:
Regarding the `default_app_lifecycle` setting:
- As of today, this parameter is not configurable as a bosh configuration parameter. It is hardcoded as `buildpack`. Can be changed, of course.
- In ccng, `default_app_lifecycle` is used to determine the default when creating an application and the lifecycle is not specified (https://v3-apidocs.cloudfoundry.org/version/3.166.0/#create-an-app).

IMHO, the specified behaviour can only be achieved if `docker` is forbidden as `default_app_lifecycle` and additionally only because the (classic) buildpack lifecycle object and the cnb lifecycle object happen to have the same structure. The CLI doesn't know the default lifecycle that shall be used for buildpack-based apps (cnb or classic buildpacks) if `--lifecycle` is not specified. The CLI just knows by cmd line parameters if is a docker or buildpack-based app.
Could work but is a bit hacky.

Alternative: the CLI has a means to query the `default_app_lifecycle` from the CF API (e.g. via info endpoint).

Reply:
I guess we can leave the details of `default_app_lifecycle` setting as implementation detail. For the RFC it is just important that the usage of classic buildpacks and CNBs as standard buildpack set should be configurable.

Reply:
During the TOC meeting we decided to wait for more clarity on this discussion before accepting the RFC.

Reply:
@beyhan @stephanme we've updated the PR with additional requirements to the default lifecycle and removed implementation details

Reply:
LGTM